110 - building package workflow

mxochicale opened the pull request:
Solves #110 

Hi all, this PR includes adds local build of the package and few other changes to improve workflow. Let me know if you have any feedback. Thanks, -- Miguel

mxochicale committed:
ignores idea path (#110)

mxochicale committed:
few improvements to the workflow to build package; sorted out LookupError issue; ignore default project name (#110)

mxochicale committed:
adds pre-commit dependencie and locally PASS pre-commit ✨ 🍰 ✨ (#110)

p-j-smith commented on README.md:
````diff
@@ -19,10 +19,14 @@ We've turned on [discussions](https://github.com/UCL-ARC/python-tooling/discussi
-   ```
-   pip install cookieninja
-   ```
+1.  Install [cookieninja](https://libraries.io/pypi/cookieninja) in a `conda`, `mamba` or `venv` environment (commented lines for conda example).
````
Should we recommend people install `cookieninja` using [`pipx`](https://pypa.github.io/pipx/) - this way they don't have tocreate an environment just for it (or add it to their project environment)

paddyroddy replied:
I think we should stick with the initial `pip install cookieninja`. For those that know, they'll know what to do anyway. Also, `conda` and `mamba` are the same.

p-j-smith commented on README.md:
````diff
@@ -19,10 +19,14 @@ We've turned on [discussions](https://github.com/UCL-ARC/python-tooling/discussi
+```
+# conda create -n ptoolingVE python=3.11 pip -c conda-forge
+# conda activate ptoolingVE
+pip install cookieninja build twine setuptools-scm pytest pre-commit
````
Once they've created their project, they should be able to do `python -m pip install -e ".[dev]"` rather than installing them manually (as these packages are listed as [optional dependencies](https://github.com/UCL-ARC/python-tooling/blob/cb5e1cbd948c9c0276a901b2eaaa9b17208b1021/%7B%7Bcookiecutter.project_slug%7D%7D/pyproject.toml#L33-L41))

paddyroddy replied:
Can we not pin the python version? The more we pin, the more we have to change later.

Agree regarding `dev`.

You also don't need to install `setuptools-scm`. Plus, we are using `tox` so no need to install `pytest`.

p-j-smith commented on README.md:
```diff
@@ -82,14 +86,17 @@ We've turned on [discussions](https://github.com/UCL-ARC/python-tooling/discussi
    git init
-   # conda create -n PROJECT_SLUG-dev python=3.11
-   # conda activate PROJECT_SLUG-dev
    python -m pip install -e .
```
as above, I think this should be:
```
python -m pip install -e ".[dev]"
```

p-j-smith commented on README.md:
````diff
@@ -82,14 +86,17 @@ We've turned on [discussions](https://github.com/UCL-ARC/python-tooling/discussi
+6. Build your package
+   ```
+   python -m build
+   python -m setuptools_scm
````
does this need to be run explicitly? I think it might be handled by `build`

paddyroddy replied:
Agreed, that second line is not needed. `build` uses the backend specified.
```suggestion
```

p-j-smith commented on README.md:
```diff
@@ -113,6 +120,32 @@ To contribute:
+  # conda create -n ptoolingVE python=3.11 pip -c conda-forge
+  # conda activate ptoolingVE
+  # conda deactivate ptoolingVE  ## TO_REMOVE
+  # conda remove -n ptoolingVE --all   ## TO_REMOVE
```
should these lines be uncommented?

p-j-smith commented on README.md:
````diff
@@ -113,6 +120,32 @@ To contribute:
+  ```
+  # conda create -n ptoolingVE python=3.11 pip -c conda-forge
+  # conda activate ptoolingVE
+  # conda deactivate ptoolingVE  ## TO_REMOVE
````
maybe change the comment from `## TO_REMOVE` to `## deactivate`, or remove the comment

p-j-smith commented on README.md:
````diff
@@ -113,6 +120,32 @@ To contribute:
+- To create and remove your virtual environment
+
+  ```
+  # conda create -n ptoolingVE python=3.11 pip -c conda-forge
````
I think they will also need `tox` in their development environment to run the tests, or `pytest` if they want to run it manually

p-j-smith commented on README.md:
````diff
@@ -125,7 +158,16 @@ To contribute:
+- To commit changes
+
+  ```
+  pre-commit run --all-files --color always --verbose
````
I think they will need to install the pre-commit hooks:
```
pre-commit install
```
before their first commit (this only needs to be done once). They shouldn't need to do:
```
pre-commit run --all-files --color always --verbose
```
unless they have commit changes to files before installing the hooks

paddyroddy replied:
Agree with Paul, but the following would be a simpler recommendation
```suggestion
  pre-commit run -a
```

paddyroddy commented on .gitignore:
```diff
@@ -1,2 +1,6 @@
+*.idea
+
+#default project path
+**/python-template
```
Just need this I think?
```suggestion
python-template
```

paddyroddy commented on README.md:
````diff
@@ -82,14 +86,17 @@ We've turned on [discussions](https://github.com/UCL-ARC/python-tooling/discussi
+   ```
+   python -m build
+   python -m setuptools_scm
+   twine check dist/*
````
Also don't need `twine`. See https://github.com/paddyroddy/python-template/blob/main/.github/workflows/deploy.yml

paddyroddy commented on README.md:
```diff
@@ -113,6 +120,32 @@ To contribute:
 
+- Clone repository
+
+  - Generate your SSH keys as suggested [here](https://docs.github.com/en/github/authenticating-to-github/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent)
```
If we're really recommending this kind of thing (which I think is beyond the scope of this project) then we could recommend the GitHub CLI instead as it is simpler https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account?tool=cli

paddyroddy commented on README.md:
````diff
@@ -113,6 +120,32 @@ To contribute:
+  - Clone the repository by typing (or copying) the following line in a terminal at your selected path in your machine:
+
+  ```
+  cd && mkdir -p $HOME/repositories/ && cd  $HOME/repositories/
````
I don't think we should be telling people where to put their `git` repos

mxochicale committed:
cleaner workflow based on p-j-smith and paddyroddy feedback and few other minor changes (#110)

mxochicale commented:
Thanks @p-j-smith  and @paddyroddy for your feedback, I think I have addressed your comments in the previous commit. Let me know if you have further comments. Thanks, --Miguel

paddyroddy committed:
Sort dependencies

mxochicale commented:
Thanks @paddyroddy  for approving this one. Let me know @p-j-smith  if you have further comments.  Hence, going for automatic merging. Thanks, --MIguel

p-j-smith reviewed (approved):
sorry for being so slow to reply! Looks good 👍
